File: java/client/src/org/openqa/selenium/interactions/touch/TouchActions.java
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.openqa.selenium.interactions.touch;

import static org.openqa.selenium.interactions.PointerInput.Kind.TOUCH;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.interactions.HasTouchScreen;
import org.openqa.selenium.interactions.Locatable;
import org.openqa.selenium.interactions.PointerInput;
import org.openqa.selenium.interactions.TouchScreen;

/**
 * Implements actions for touch enabled devices, reusing the available composite and builder design
 * patterns from Actions.
 */
public class TouchActions extends Actions {

  private final PointerInput touchPointer = new PointerInput(TOUCH, "touch screen");
  protected TouchScreen touchScreen;

  public TouchActions(WebDriver driver) {
    super(driver);
    if (driver instanceof HasTouchScreen) {
      this.touchScreen = ((HasTouchScreen) driver).getTouch();
    } else {
      this.touchScreen = null;
    }
  }

  /**
   * Allows the execution of single tap on the screen, analogous to click using a Mouse.
   *
   * @param onElement the {@link WebElement} on the screen.
   * @return self
   */
  public TouchActions singleTap(WebElement onElement) {
    if (touchScreen != null) {
      action.addAction(new SingleTapAction(touchScreen, (Locatable) onElement));
    }
    tick(touchPointer.createPointerDown(0));
    tick(touchPointer.createPointerUp(0));
    return this;
  }

  /**
   * Allows the execution of the gesture 'down' on the screen. It is typically the first of a
   * sequence of touch gestures.
   *
   * @param x The x coordinate relative to the viewport
   * @param y The y coordinate relative to the viewport
   * @return self
   */
  public TouchActions down(int x, int y) {
    if (touchScreen != null) {
      action.addAction(new DownAction(touchScreen, x, y));
    }
    return this;
  }

  /**
   * Allows the execution of the gesture 'up' on the screen. It is typically the last of a sequence
   * of touch gestures.
   *
   * @param x The x coordinate relative to the viewport
   * @param y The y coordinate relative to the viewport
   * @return self
   */
  public TouchActions up(int x, int y) {
    if (touchScreen != null) {
      action.addAction(new UpAction(touchScreen, x, y));
    }
    return this;
  }

  /**
   * Allows the execution of the gesture 'move' on the screen.
   *
   * @param x The x coordinate relative to the viewport
   * @param y The y coordinate relative to the viewport
   * @return self
   */
  public TouchActions move(int x, int y) {
    if (touchScreen != null) {
      action.addAction(new MoveAction(touchScreen, x, y));
    }
    return this;
  }

  /**
   * Creates a scroll gesture that starts on a particular screen location.
   *
   * @param onElement the {@link WebElement} where the scroll starts.
   * @param xOffset   The x offset to scroll
   * @param yOffset   The y offset to scroll
   * @return self
   */
  public TouchActions scroll(WebElement onElement, int xOffset, int yOffset) {
    if (touchScreen != null) {
      action.addAction(new ScrollAction(touchScreen, (Locatable) onElement, xOffset, yOffset));
    }
    return this;
  }

  /**
   * Allows the execution of double tap on the screen, analogous to double click using a Mouse.
   *
   * @param onElement The {@link WebElement} to double tap
   * @return self
   */

  public TouchActions doubleTap(WebElement onElement) {
    if (touchScreen != null) {
      action.addAction(new DoubleTapAction(touchScreen, (Locatable) onElement));
    }
    return this;
  }

  /**
   * Allows the execution of long press gestures.
   *
   * @param onElement The {@link WebElement} to long press
   * @return self
   */

  public TouchActions longPress(WebElement onElement) {
    if (touchScreen != null) {
      action.addAction(new LongPressAction(touchScreen, (Locatable) onElement));
    }
    return this;
  }

  /**
   * Allows the view to be scrolled by an x and y offset.
   *
   * @param xOffset The horizontal offset relative to the viewport
   * @param yOffset The vertical offset relative to the viewport
   * @return self
   */

  public TouchActions scroll(int xOffset, int yOffset) {
    if (touchScreen != null) {
      action.addAction(new ScrollAction(touchScreen, xOffset, yOffset));
    }
    return this;
  }

  /**
   * Sends a flick gesture to the current view.
   *
   * @param xSpeed The horizontal speed in pixels/second
   * @param ySpeed The vertical speed in pixels/second
   * @return self
   */

  public TouchActions flick(int xSpeed, int ySpeed) {
    if (touchScreen != null) {
      action.addAction(new FlickAction(touchScreen, xSpeed, ySpeed));
    }
    return this;
  }

  /**
   * Allows the execution of flick gestures starting in a location's element.
   *
   * @param onElement The {@link WebElement} to flick on
   * @param xOffset   The x offset relative to the viewport
   * @param yOffset   The y offset relative to the viewport
   * @param speed speed to flick, 0 = normal, 1 = fast, 2 = slow
   * @return self
   */

  public TouchActions flick(WebElement onElement, int xOffset, int yOffset, int speed) {
    if (touchScreen != null) {
      action.addAction(new FlickAction(touchScreen, (Locatable) onElement, xOffset, yOffset, speed));
    }
    return this;
  }
}

File: dotnet/src/webdriver/Interactions/TouchActions.cs
// <copyright file="TouchActions.cs" company="WebDriver Committers">
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The SFC licenses this file
// to you under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>

using System;
using OpenQA.Selenium.Internal;

namespace OpenQA.Selenium.Interactions
{
    /// <summary>
    /// Provides a mechanism for building advanced interactions with the browser.
    /// </summary>
    public class TouchActions : Actions
    {
        private readonly TimeSpan DefaultScrollMoveDuration = TimeSpan.FromSeconds(2);
        private readonly TimeSpan DefaultLongPressDuration = TimeSpan.FromSeconds(2);
        private readonly TimeSpan DefaultFlickDuration = TimeSpan.FromMilliseconds(100);
        private ActionBuilder actionBuilder = new ActionBuilder();
        private PointerInputDevice defaultTouchscreen = new PointerInputDevice(PointerKind.Touch, "default touch pointer");

        /// <summary>
        /// Initializes a new instance of the <see cref="TouchActions"/> class.
        /// </summary>
        /// <param name="driver">The <see cref="IWebDriver"/> object on which the actions built will be performed.</param>
        public TouchActions(IWebDriver driver)
            : base(driver)
        {
        }

        /// <summary>
        /// Taps the touch screen on the specified element.
        /// </summary>
        /// <param name="onElement">The element on which to tap.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions SingleTap(IWebElement onElement)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(onElement, 0, 0, TimeSpan.Zero));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Presses down at the specified location on the screen.
        /// </summary>
        /// <param name="locationX">The x coordinate relative to the view port.</param>
        /// <param name="locationY">The y coordinate relative to the view port.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions Down(int locationX, int locationY)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(CoordinateOrigin.Viewport, locationX, locationY, TimeSpan.Zero));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Releases a press at the specified location on the screen.
        /// </summary>
        /// <param name="locationX">The x coordinate relative to the view port.</param>
        /// <param name="locationY">The y coordinate relative to the view port.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions Up(int locationX, int locationY)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(CoordinateOrigin.Viewport, locationX, locationY, TimeSpan.Zero));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Moves to the specified location on the screen.
        /// </summary>
        /// <param name="locationX">The x coordinate relative to the view port.</param>
        /// <param name="locationY">The y coordinate relative to the view port.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions Move(int locationX, int locationY)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(CoordinateOrigin.Viewport, locationX, locationY, TimeSpan.Zero));
            return this;
        }

        /// <summary>
        /// Scrolls the touch screen beginning at the specified element.
        /// </summary>
        /// <param name="onElement">The element on which to begin scrolling.</param>
        /// <param name="offsetX">The x coordinate relative to the view port.</param>
        /// <param name="offsetY">The y coordinate relative to the view port.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions Scroll(IWebElement onElement, int offsetX, int offsetY)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(onElement, 0, 0, TimeSpan.Zero));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(CoordinateOrigin.Viewport, offsetX, offsetX, DefaultScrollMoveDuration));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Double-taps the touch screen on the specified element.
        /// </summary>
        /// <param name="onElement">The element on which to double-tap.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions DoubleTap(IWebElement onElement)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(onElement, 0, 0, TimeSpan.Zero));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Presses and holds on the touch screen on the specified element.
        /// </summary>
        /// <param name="onElement">The element on which to press and hold</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions LongPress(IWebElement onElement)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(onElement, 0, 0, TimeSpan.Zero));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePause(DefaultLongPressDuration));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Scrolls the touch screen to the specified offset.
        /// </summary>
        /// <param name="offsetX">The horizontal offset relative to the view port.</param>
        /// <param name="offsetY">The vertical offset relative to the view port.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions Scroll(int offsetX, int offsetY)
        {
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(CoordinateOrigin.Viewport, offsetX, offsetX, DefaultScrollMoveDuration));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Flicks the current view.
        /// </summary>
        /// <param name="speedX">The horizontal speed in pixels per second.</param>
        /// <param name="speedY">The vertical speed in pixels per second.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions Flick(int speedX, int speedY)
        {
            int offsetX = Convert.ToInt32(Math.Round(speedX * DefaultFlickDuration.TotalSeconds, 0));
            int offsetY = Convert.ToInt32(Math.Round(speedY * DefaultFlickDuration.TotalSeconds, 0));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(CoordinateOrigin.Viewport, offsetX, offsetX, DefaultFlickDuration));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }

        /// <summary>
        /// Flicks the current view starting at a specific location.
        /// </summary>
        /// <param name="onElement">The element at which to start the flick.</param>
        /// <param name="offsetX">The x offset relative to the viewport.</param>
        /// <param name="offsetY">The y offset relative to the viewport.</param>
        /// <param name="speed">The speed in pixels per second.</param>
        /// <returns>A self-reference to this <see cref="TouchActions"/>.</returns>
        public TouchActions Flick(IWebElement onElement, int offsetX, int offsetY, int speed)
        {
            int normalizedOffsetX = Convert.ToInt32(offsetX / speed * DefaultFlickDuration.TotalSeconds);
            int normalizedOffsetY = Convert.ToInt32(offsetY / speed * DefaultFlickDuration.TotalSeconds);
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(onElement, 0, 0, TimeSpan.Zero));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerDown(MouseButton.Touch));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerMove(CoordinateOrigin.Viewport, normalizedOffsetX, normalizedOffsetY, DefaultFlickDuration));
            this.actionBuilder.AddAction(this.defaultTouchscreen.CreatePointerUp(MouseButton.Touch));
            return this;
        }
    }
}
